DO NOT MERGE: MVVM prototyping

garrettjonesgoogle opened the pull request:
In this prototype, the surface data classes only contain the data that varies
between APIs; otherwise, all other code is in the template. This results
in a much more manageable structure; the template becomes simpler while
still remaining direct.

Note that I used appalling coding practices here in order to prototype
faster (e.g. all my surface data classes just have public fields). I'm
more interested in feedback on the general approach and flow of data.

garrettjonesgoogle commented:
I have continued developing the "surface data" approach and I think it is working out nicely. In comparison to the existing code:

* The context will go away completely. Most of the functionality ends up in the transformer or ModelToJavaTypeTable.
* The template will be shorter, and it won't do any naming conversions. It is much more straightforward now.
* There are a lot of view model / surface data classes, currently only with public fields. These can probably all be converted to AutoValue classes.

As for testing, I think what I want to do is turn on code coverage, and just add unit tests for code paths that aren't currently covered by the baselines. We can add more coverage as time goes on.

bjwatson commented on src/test/java/com/google/api/codegen/DiscoveryGeneratorTestBase.java:
```diff
@@ -56,22 +56,22 @@ public DiscoveryGeneratorTestBase(
-    if (config == null) {
-      return;
-    }
+    //    try {
```
Add a comment why this is commented out, so we can circle back and fix it.

garrettjonesgoogle replied:
Done.

bjwatson commented on src/main/java/com/google/api/codegen/gapic/SurfaceGapicProvider.java:
```diff
@@ -0,0 +1,177 @@
+
+import javax.annotation.Nullable;
+
+public class SurfaceGapicProvider implements GapicProvider<Interface> {
```
This currently is has a lot of Java-specific references for appearing to be language-agnostic code.

garrettjonesgoogle replied:
True, but this part is just hacked up to glue everything together, and will be made generic at a later point in the refactoring.

garrettjonesgoogle commented:
Ok I have added support for generating most of the PHP surface (except sample code), and in the process I have refactored a lot of the transformation code to be language-independent. I think this is a good checkpoint for people to look at where this is going before I do the last stretch of fully supporting all of the Java/PHP codegen, polishing the coding practices (e.g. removing public fields), and deleting out the existing codegen for Java/PHP.

shinfan commented on src/main/java/com/google/api/codegen/transformer/ModelToSurfaceContext.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Collection;
+
+@AutoValue
+public abstract class ModelToSurfaceContext {
```
I feel the functionality of this class is very general. Will there be more contexts added in the future (such as ModelToXContext)?
If not, we can just call this TransformerContext.

garrettjonesgoogle replied:
I was thinking more along the lines that more types of transformations might be done, but not necessarily ModelToX, but some other AToB transformation, which might need its own context.

shinfan commented on src/main/java/com/google/api/codegen/transformer/IdentifierNamer.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * An instance of IdentifierNamer provides language-specific names or other strings.
+ */
+public interface IdentifierNamer {
```
IdentifierTransformer sounds like a better name to me

garrettjonesgoogle replied:
But it's not transforming identifiers, it's creating them - i.e. identifiers are the output, not the input.

shinfan commented on src/main/java/com/google/api/codegen/transformer/ModelToJavaSurfaceTransformer.java:
```diff
@@ -0,0 +1,256 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.transformer;
```
It may worth considering separating this package into sub-packages by language (e.g. com.google.api.codegen.transformer.java).

garrettjonesgoogle replied:
Alternatively I was thinking of just moving the language-specific classes into the already-existing language-specific packages.

shinfan replied:
I actually like the idea of packaging classes based on their concepts (surface, transformers. etc.) on the top level which makes the architecture easier to see.

bjwatson replied:
I agree with @shinfan. I like making the architecture obvious via the package structure.

garrettjonesgoogle replied:
I think this means orienting away from the top-level language directories and towards top-level architecture directories. That makes sense given that we are trying to make more of the toolkit language-independent.

bjwatson replied:
Yeah, definitely.

garrettjonesgoogle commented:
Sorry about the mess created by this latest push. I was trying to pull in changes from master but it horked up the commit history.

shinfan commented on src/main/java/com/google/api/codegen/transformer/ModelToSurfaceTransformer.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.List;
+
+public interface ModelToSurfaceTransformer {
```
Rename to ModelToViewTransformer?

garrettjonesgoogle replied:
Will do.

shinfan commented on src/main/java/com/google/api/codegen/transformer/JavaSurfaceNamer.java:
```diff
@@ -0,0 +1,227 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class JavaSurfaceNamer implements SurfaceNamer {
```
As discussed we should probably avoid using "surface".
I think instead of calling it "Xnamer", we can call it "JavaNameProvider"

garrettjonesgoogle replied:
In this case I specifically chose "Surface" because the names in this interface are for the surface. For your test generation, you'll probably want your own namer.

shinfan replied:
Most of the names are the same between tests and the clients (e.g. variable name, method name, class name) so it would be good if we can share some of the functionalities for both cases.

garrettjonesgoogle replied:
I refactored the naming stuff a bit; there should be some reusable stuff for your test work now.

bjwatson commented on src/main/java/com/google/api/codegen/gapic/SurfaceGapicProvider.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.Map;
+import java.util.TreeMap;
+
+public class SurfaceGapicProvider implements GapicProvider<Interface> {
```
I'd be tempted to make this an `@AutoValue` class to avoid the `Builder` boilerplate.

Also, consider making this class `final`, since it does not seem intended to be overridden.

garrettjonesgoogle replied:
I fought with it a little bit but gave up. The issue is that I don't want getters for the fields in the main class, because they are only used internally, but using AutoValue for the builder forces you to make getters for them in the main class.

bjwatson replied:
They can be package private: https://github.com/google/auto/blob/master/value/userguide/howto.md#private_accessors

I guess it's not a big deal, since you probable don't care about `equals`, `hashCode`, or `toString`. This is not really a value class.

shinfan commented on src/main/java/com/google/api/codegen/util/LanguageNamer.java:
```diff
@@ -0,0 +1,32 @@
+
+public interface LanguageNamer {
+
+  String getClassName(Name name);
```
Can these methods be static?

bjwatson commented on src/main/java/com/google/api/codegen/gapic/MainGapicProviderFactory.java:
```diff
@@ -142,6 +149,24 @@
+              .setModel(model)
+              .setSnippetSetRunner(
+                  new CommonSnippetSetRunner(
+                      SnippetSetRunner.SNIPPET_RESOURCE_ROOT, new JavaSnippetUtil()))
```
The `resourceRoot` argument seems useless if it's always `SNIPPET_RESOURCE_ROOT`.  Why define `resourceRoot` in `CommonSnippetSetRunner` anyway, if `ModelToSurfaceTransformer` defines the resource files?

garrettjonesgoogle replied:
Hmm yeah it doesn't make sense to pass the resource root in here any more. Maybe it could be provided by ViewModel instead.

bjwatson commented on src/main/java/com/google/api/codegen/viewmodel/CommonSnippetSetRunner.java:
```diff
@@ -0,0 +1,49 @@
+  private String resourceRoot;
+  private Object utilObject;
+
+  public CommonSnippetSetRunner(String resourceRoot, Object utilObject) {
```
Why define `resourceRoot` here, if `ModelToSurfaceTransformer` defines the resource files anyway? It seems like we could eliminate `resourceRoot`, and possibly move `utilObject` into the transformer object, leaving this as just a utility class.

garrettjonesgoogle replied:
But utilObject still needs to be passed to the snippet runtime somehow.

bjwatson commented on src/main/java/com/google/api/codegen/gapic/SurfaceGapicProvider.java:
```diff
@@ -0,0 +1,111 @@
+      return null;
+    }
+
+    List<ViewModel> surfaceDocs = modelToViewTransformer.transform(model);
```
The following code could be more efficient if it passed `snippetFileName` into the `transform()` method, rather than performing all transformations and then filtering them.

garrettjonesgoogle replied:
This feature is only used by tests, and it doesn't seem to affect how long the tests take to run. If it's ever an issue, I'd rather cache the result of the full transformation and reuse it.

bjwatson replied:
Ok, that's a fair approach.

bjwatson commented on src/test/java/com/google/api/codegen/JavaSurfaceCodeGeneratorTest.java:
```diff
@@ -0,0 +1,56 @@
+  public JavaSurfaceCodeGeneratorTest(
+      String name, String idForFactory, String[] gapicConfigFileNames, String snippetName) {
+    super(name, idForFactory, gapicConfigFileNames, snippetName);
+    getTestDataLocator().addTestDataSource(com.google.api.codegen.java.JavaTypeTable.class, "");
```
Specifying the test data path this way is non-obvious, and has nothing to do with `JavaTypeTable`. I know you inherited this from the existing code, but maybe we could start moving away from this pattern?

garrettjonesgoogle replied:
I'll keep it in mind.

bjwatson commented on src/main/java/com/google/api/codegen/java/JavaTypeTable.java:
```diff
@@ -0,0 +1,126 @@
+import java.util.List;
+import java.util.Map;
+
+public class JavaTypeTable {
```
Can be package private and final

garrettjonesgoogle replied:
Can't be package private, it is used outside of this package (.../transformer)

bjwatson commented on src/main/java/com/google/api/codegen/php/PhpTypeTable.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Collections;
+import java.util.List;
+
+public class PhpTypeTable {
```
Can be package private and final

garrettjonesgoogle replied:
Can't be package private, it is used outside of this package (.../transformer)

bjwatson commented on src/main/java/com/google/api/codegen/transformer/JavaSurfaceNamer.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class JavaSurfaceNamer extends SurfaceNamer {
```
Can be package private and final

bjwatson commented on src/main/java/com/google/api/codegen/transformer/ModelToJavaTypeTable.java:
```diff
@@ -0,0 +1,255 @@
+import java.io.File;
+import java.util.List;
+
+public class ModelToJavaTypeTable implements ModelTypeTable {
```
Can be package private and final

bjwatson commented on src/main/java/com/google/api/codegen/transformer/ModelToJavaTypeTable.java:
```diff
@@ -0,0 +1,255 @@
+import java.util.List;
+
+public class ModelToJavaTypeTable implements ModelTypeTable {
+  private JavaTypeTable javaTypeTable;
```
The fact that this holds mutable state is not obvious, nor is it obvious that the state is a list of import statements. Also, the decomposition into `ModelToJavaTypeTable` and `JavaTypeTable` is not obvious.

Similar thoughts apply to the PHP code.

garrettjonesgoogle replied:
The function names should make it clear that it holds mutable state (e.g. getAndSaveX). I used the name "TypeTable" to hint at similarities with a symbol table. 

What do you mean by decomposition here?

bjwatson commented on src/main/java/com/google/api/codegen/transformer/ModelToPhpTypeTable.java:
```diff
@@ -0,0 +1,205 @@
+
+import org.apache.commons.lang3.NotImplementedException;
+
+public class ModelToPhpTypeTable implements ModelTypeTable {
```
Can be package private and final

bjwatson commented on src/main/java/com/google/api/codegen/transformer/PhpSurfaceNamer.java:
```diff
@@ -0,0 +1,84 @@
+import com.google.api.tools.framework.model.Method;
+import com.google.api.tools.framework.model.TypeRef;
+
+public class PhpSurfaceNamer extends SurfaceNamer {
```
Can be package private and final

garrettjonesgoogle replied:
I don't think it's quite the right stage to go around making everything package-private and final that can be. I think that makes sense when toolkit becomes a library that other libraries depend on, but we're not there yet.

bjwatson commented on src/main/java/com/google/api/codegen/java/JavaTypeTable.java:
```diff
@@ -0,0 +1,126 @@
+  /**
+   * A map from unboxed Java primitive type name to boxed counterpart.
+   */
+  private static final ImmutableMap<String, String> BOXED_TYPE_MAP =
```
Define constants for these, especially the keys which are also used in `ModelToJavaType.PRIMITIVE_TYPE_MAP`.

garrettjonesgoogle replied:
That seems like needless overhead...

bjwatson commented on src/main/java/com/google/api/codegen/transformer/ModelToJavaSurfaceTransformer.java:
```diff
@@ -0,0 +1,255 @@
+    return surfaceDocs;
+  }
+
+  public List<ViewModel> transform(Interface service) {
```
This should call two private methods, one for each `templateFileName`. That would make the code a little more readable.

garrettjonesgoogle replied:
I'll do that.

bjwatson commented on src/main/java/com/google/api/codegen/transformer/ModelToJavaSurfaceTransformer.java:
```diff
@@ -0,0 +1,255 @@
+  public List<ViewModel> transform(Interface service) {
+    SurfaceTransformerContext context =
+        SurfaceTransformerContext.create(
+            service, cachedApiConfig, new ModelToJavaTypeTable(), new JavaSurfaceNamer());
```
The context doesn't really help much, since it's not shared outside this class (same for the PHP transformer). Also, it could too rigidly constrains what a transformer needs for a language.

Note that `cachedApiConfig` is globally available anyway, and `JavaSurfaceNamer` is stateless, so it could also be globally available.

garrettjonesgoogle replied:
It's passed all over the place to the other transformers:

* ApiMethodTransformer
* InitCodeTransformer
* PageStreamingTransformer
* PathTemplateTransformer

It's not meant to be a transformer context for all transformers; e.g. Shin's test transformers would use their own context. It is as specific as it needs to be.

I don't think we should be caching api config, since we want to support batch code generation over multiple APIs later. Also we shouldn't make the surface namer into a singleton either, if we want to support batch code generation over multiple languages.

garrettjonesgoogle replied:
Ok I realized my remark about caching api config doesn't quite make sense - I'm the one who named the variable `cachedApiConfig`. In any case, I think that architecturally it might make more sense for it to be passed in at the same time as the model, in the transform() function.

bjwatson commented on src/main/java/com/google/api/codegen/transformer/ModelToJavaTypeTable.java:
```diff
@@ -0,0 +1,255 @@
+   * A map from primitive types in proto to Java counterparts.
+   */
+  private static final ImmutableMap<Type, String> PRIMITIVE_TYPE_MAP =
+      ImmutableMap.<Type, String>builder()
```
These type strings should be defined as constants somewhere.

garrettjonesgoogle replied:
That seems like needless overhead...

shinfan commented on src/main/java/com/google/api/codegen/transformer/MethodTransformerContext.java:
```diff
@@ -0,0 +1,63 @@
+      ModelTypeTable typeTable,
+      SurfaceNamer namer,
+      Method method,
+      MethodConfig methodConfig) {
```
methodConfig seems redundant here. We can just use apiConfig.getInterfaceConfig(interfaze).getMethodConfig(method).

garrettjonesgoogle replied:
So this is only called by SurfaceTransformerContext, which already does that in its own constructor, and just passes it on to this class.

garrettjonesgoogle commented:
I have split this apart:

* PR against master for PHP: https://github.com/googleapis/toolkit/pull/317
* Java support, diffbased on the above: https://github.com/garrettjonesgoogle/toolkit/pull/1

Closing out the present PR.

bjwatson commented on src/test/java/com/google/api/codegen/testdata/php_main_library.baseline:
```diff
@@ -230,6 +228,7 @@ class LibraryServiceApi
         return self::$archivedBookNameTemplate;
     }
 
+
```
Can we avoid the extra blank line, or is this an idiomatic PHP thing?

garrettjonesgoogle replied:
FYI you are commenting on the closed PR here.

The snippet language is really finicky with newlines - it's quite hard to get it exactly right.

bjwatson replied:
Oops. I have too many tabs open.

That sounds like a bug in the snippet language, since the whole point of it is to get formatting right. I strongly prefer to avoid extra blank lines unless it's a thing that a particular language community likes to do.

bjwatson commented on src/test/java/com/google/api/codegen/testdata/php_main_library.baseline:
```diff
@@ -439,15 +438,17 @@ class LibraryServiceApi
+     *     @var SomeMessage $message
+     *          Field to verify that message-type query parameter gets flattened.
      *     @var StringBuilder $stringBuilder
+     *
```
Looks like another extra blank line.
